rascal-lsp/src/main/rascal/lang/rascal/lsp/Outline.rsc

```diff
@@ -31,38 +31,58 @@ import String;
 import ParseTree;
 import lang::rascal::\syntax::Rascal;
 import util::LanguageServer;
+import util::ErrorRecovery;
 
 list[DocumentSymbol] outlineRascalModule(start[Module] \mod) {
     m= \mod.top;
-    children = [];
-
-    top-down-break visit (m) {
-        case (Declaration) `<Tags _> <Visibility _> <Type t> <{Variable ","}+ vars>;`:
-            children += [symbol(clean("<v.name>"), variable(), v@\loc, detail="variable <t> <v>") | v <- vars];
-
-        case (Declaration) `<Tags _> <Visibility _> anno <Type t> <Type ot>@<Name name>;`:
-            children +=  [symbol(clean("<name>"), field(), t@\loc, detail="anno <t> <ot>")];
 
-        case (Declaration) `<Tags _> <Visibility _> alias <UserType u> = <Type al>;`:
-            children += [symbol(clean("<u.name>"), struct(), u@\loc, detail="<u> = <al>")];
+    if (!(m has header) || hasErrors(m.header)) {
+        return [];
+    }
 
-        case (Declaration) `<Tags _> <Visibility _> tag <Kind k> <Name name> on <{Type ","}+ ts>;`:
-            children += [symbol(clean("<name>"), \key(), name@\loc, detail="tag <k> <name> on <ts>")];
+    children = [];
 
-        case (Declaration) `<Tags _> <Visibility _> data <UserType u> <CommonKeywordParameters kws>;`: {
-            kwlist = [symbol(".<k.name>", property(), k@\loc, detail="<k.\type>") | kws is present, KeywordFormal k <- kws.keywordFormalList];
-            children += [symbol("<u.name>", struct(), u@\loc, detail="data <u> <kws>", children=kwlist)];
+    top-down-break visit (m) {
+        case decl: (Declaration) `<Tags _> <Visibility _> <Type t> <{Variable ","}+ vars>;`:
+            if (!hasErrors(decl)) {
+                children += [symbol(clean("<v.name>"), variable(), v@\loc, detail="variable <t> <v>") | v <- vars, !hasErrors(v)];
+            }
+
+        case decl: (Declaration) `<Tags _> <Visibility _> anno <Type t> <Type ot>@<Name name>;`:
+            if (!hasErrors(decl)) {
+                children +=  [symbol(clean("<name>"), field(), t@\loc, detail="anno <t> <ot>")];
+            }
+
+        case decl: (Declaration) `<Tags _> <Visibility _> alias <UserType u> = <Type al>;`:
+            if (!hasErrors(decl)) {
+                children += [symbol(clean("<u.name>"), struct(), u@\loc, detail="<u> = <al>")];
```

Comment:
but here we use field projecten `u.name` and `u` might not even have a name if `UserType u` is an error tree. This is going to happen all over the place for error trees, so I'd like a more reusable/generic solution then having to guard all the field projections. It will become a cross-cutting and scattered concern everywhere otherwise, while error trees are a builtin language concept that features like `.` projection {c,sh,w}ould handle gracefully.

Reply:
I think we may assume that if a programmer uses `u.name` then in that place `u.name` normally does not fail, and so it should also not fail for error children.

* if u is an error tree, and the error "dot" is after the `name` field, then field projection can still work as it was supposed to.
* otherwise if the "dot" is before the `name` field or on it, then it may be possible to return an empty stub error tree for `name` instead of failing.

If we have such a semantics for field projection, there would be a lot less pressure on user code for dealing with all kinds of exceptions. In general for productions with `n` children, there are I guess something like `n!/(2 * (n-2)!)`  possible combinations of having and not having error children, so that can become untenable.

Another option is to rewrite all field projections to pattern matches, which would fail in the absence of a field. But in this example that would lead to a lot of superfluous case distintions:

```rascal
syntax UserType
	= name: QualifiedName name 
	| parametric: QualifiedName name >> "[" "[" {Type ","}+ parameters "]" ;
```

Here we really do not care about the difference between the two rules but rather their commonality which is fine by the `.name` field projection. So I guess pattern matching is not a solution here.